### ISSUE TITLE
JUNE: API Refactor

### ISSUE THREAD

**Issue by NealHumphrey** [codefordc/housing-insights#249]
Turn our current one-page, raw SQL API approach into something scalable:
- [ ] Create a logical separation of concerns among different API endpoint types, with an appropriate 
- [ ] Make sure we have a consistent approach to how endpoints are served, to provide an easier method for the front end to switch between data sources and add new graphs. Coordinate with the filter team ( #248) re: filter endpoint.
- [ ] Potentially use the SQLAlchemy ORM if this provides a big enough value add. 
- [ ] Keep in mind that in V2 (after June) we will want to add a 'download data' function that serves CSV's in parallel to the JSON data. 

Any potential data ingestion process update using SQL Alchemy should wait until V2 (after June 13), if it seems worth doing a refactor. 

Target Date: May 23 to have a refactored API format, so that new endpoints can again start being added.

**Comment by NealHumphrey** [comment]
Hey @ohaaga and @jasonrhaas just wanted to check in on the status of work on this. Jason mentioned you guys talked last week. What are you thinking re: approach? What do you think about a target of next week for having a structure in place for the refactor, so that we can open the code up to adding other endpoints as needed?

As a heads up, there are two more endpoints that our front end needs - one that @Nunie123 was working on to serve the Filter team, and one that just serves up the meta.json. I'm adding these in directly but can worry about moving them into a refactored API afterwards if you already have an WIP refactor (not sure if you do).

**Comment by jasonrhaas** [comment]
Other than prototyping and planning I have not been able to dedicate much time to this yet.  I should be able to dig into this a bit more this week.  Once I did into the data ingestion more I can have a better idea of how long this should take to do.

**Comment by jasonrhaas** [comment]
@NealHumphrey @ohaaga OK I will have time to start on this starting tomorrow.  I've been working on some other projects but plan to dig into this a bit.  I need to familiarize myself with the data structure again to create the correct declarations for SQL Alchemy.

**Comment by ajhalani** [comment]
Mhoofl

**Comment by ajhalani** [comment]
Mhoofl

On May 15, 2017 6:32 PM, "Neal Humphrey" <notifications@github.com> wrote:

Hey @ohaaga <https://github.com/ohaaga> and @jasonrhaas
<https://github.com/jasonrhaas> just wanted to check in on the status of
work on this. Jason mentioned you guys talked last week. What are you
thinking re: approach? What do you think about a target of next week for
having a structure in place for the refactor, so that we can open the code
up to adding other endpoints as needed?

As a heads up, there are two more endpoints that our front end needs - one
that @Nunie123 <https://github.com/nunie123> was working on to serve the
Filter team, and one that just serves up the meta.json. I'm adding these in
directly but can worry about moving them into a refactored API afterwards
if you already have an WIP refactor (not sure if you do).

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/codefordc/housing-insights/issues/249#issuecomment-301623925>,
or mute the thread

**Comment by NealHumphrey** [comment]
@jasonrhaas Awesome! Glad to hear it. 

Here's a few additional things to consider based on our latest developments:
1) All the endpoints needed for V1 are in the current draft except for one. That's the filter data (see below). But, we definitely want any improved API structure completed for the June 13 launch because for V2 we will be adding additional endpoints for the building view (all our front end people are working on the filter/browse view so building view is on hold). 

2) We decided to do the filtering on the front end rather than via the API. This means there is just one additional endpoint needed; @Nunie123 is the one that wrote the SQL query needed for that, he was going to add the endpoint as well - for now it'll be just a quick 'run this query' thing. We will probably merge that into dev before this task is done (so JS can use it) so it'll require a small hotfix to the refactored branch. 

3) Currently we deploy the whole /python/api folder to the server; but, in V2 we'll want the server to also run weekly data collection + ingestion jobs. So it probably makes sense to move the application.py file up a level, into the /python folder so that the server can have access to the rest of the codebase. . Note, I named it application.py because that's what AWS expects to use its default Flask deploy setup. 

This could be as simple as just setting up a logical separation of what files to put differnet raw-SQL endpoints into. SQLAlchemy would definitely be cleaner though. You make the call based on how hard implementing sqlalchemy across the board will be. 

Finally, @jkwening is another good resource if you have data ingestion process questions. I'm away from a computer during most of the day now through Monday, but will be checking in daily.

**Comment by jasonrhaas** [comment]
Trying to use SQLAlchemy's [automap](http://docs.sqlalchemy.org/en/latest/orm/extensions/automap.html) feature to get this thing jump started.  Once the declarations in there, we get a whole bunch of nice built in API features that come with Flask RestLess.  Once that is up and going the next step will be to translate the endpoints on the existing API to the new format.  Hopefully this can be done nicely with built in Flask Restless and/or SQLAlchemy features.  If not we can always continue to use the raw SQL syntax like it is now.
